### PR TITLE
Rewrite library using procedural macros.

### PR DESCRIPTION
**Pull request by Kobzol**
I noticed this nice crate today and took it as an exercise to rewrite it using procedural macros. It turned out to be pretty straightforward, but it was my first experience with procedural macros, so it's possible that I made some terrible errors :)

I changed the syntax a little bit, because it was problematic to use `#[target_method(...)]`. `target_method` is an unknown attribute and that must be feature gated (although now that I think of it, you could define the attribute in this crate as a no-op attribute and then let users import it and use it).
The `use len` syntax is obviously not backwards-compatible and it is not that much obvious in the code, but it's much less verbose than the attribute, which is the whole point of this crate.

```rust
impl<T> Stack<T> {
    /// Allocate an empty stack
    pub fn new() -> Stack<T> {
        Stack { inner: vec![] }
    }

    delegate! { self.inner
        fn size(&self) -> usize; use len  // calls self.inner.len()

        /// doc comment
        fn is_empty(&self) -> bool;

        fn push(&mut self, v: T);
        fn pop(&mut self) -> Option<T>;
        fn peek(&self) -> Option<&T>; use last
        fn clear(&mut self);
        fn into_boxed_slice(self) -> Box<[T]>;
    }
}
```

I parse the functions as function declarations within a Trait, so they must have a trailing semicolon. Ideally I would like to write
```rust
delegate! { self.inner 
    fun1, fun2, fun3
}
```
however AFAIK that would require a compiler plugin to find the corresponding signatures of the delegated functions from the delegated struct.

I think that it's much better for this crate to use procedural macros than to reimplement the syntax parser by hand using classic macros. Please let me know what you think :)

BTW this fixes https://github.com/chancancode/rust-delegate/issues/5 :)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,1936 +1,120 @@
...
+#[proc_macro]
+pub fn delegate(tokens: TokenStream) -> TokenStream {
+    let delegator: DelegatedSegment = syn::parse_macro_input!(tokens);
+
+    let delegator_attribute = delegator.delegator;
+    let functions = delegator.methods.iter().map(|fun| {
+        let input = &fun.method;
+        let signature = &input.sig;
+        let inputs = &input.sig.decl.inputs;
+        let attrs = &input.attrs;
```
Comment (gavento) [inline src/lib.rs]: Maybe add `#[inline(always)]` to `attrs` (fancy customization: unless it already contains `#[inline(never)]`.

**Comment by Kobzol** [comment]
I changed the syntax, now it should be compatible with the old style.

**Comment by Boscop** [comment]
@Kobzol Does your rewritten version also support multiple targets in one macro call, and can `target` be a member of a member (of a member)?

**Comment by Boscop** [comment]
Btw, I'm looking forward to using the proc-macro version :)

**Comment by Boscop** [comment]
@Kobzol I just tried your fork. Wow, it supports delegation to members of members. Please merge this :)

**Comment by chancancode** [comment]
@kobzol @boscop any chance that either of you would be happy to take over the repo/create? 😅

**Comment by Boscop** [comment]
I'd be interested if @Kobzol is not interested. I use this crate a lot, and consider it one of my go-to convenience crates in the ecosystem..
Kobzol did the work with the proc-macro rewrite, so it's his call, if he wants to take over :)

**Comment by chancancode** [comment]
> I use this crate a lot, and consider it one of my go-to convenience crates in the ecosystem..

That's awesome 😄

I just don't happen to be writing a lot of Rust recently so I think it would be great for someone who is using this more to own and maintain it. I guess we'll give @Kobzol a couple of days? (But it also seems fine for both of you to have commit/publish, something something bus factor.)

**Comment by Kobzol** [comment]
I'd be happy to maintain this crate. As you said, the more maintainers the better, so it would be nice if you gave commit rights to both me and @Boscop if you want to pass the ownership along. The first step after that would be revising and integrating the `proc-macro` rewrite, I'd take care of that.

**Comment by chancancode** [comment]
Awesome! Thank you very much for stepping up, invites are sent! I was going to transfer the repo to @Kobzol too but you already have a fork with the same name so GitHub won’t do it 🤷‍♂ If that becomes important in the future for some reason, let me know and we can try again.

**Comment by chancancode** [comment]
Also feel free to change the  cargo author metadata!

**Comment by Boscop** [comment]
Thanks for the invite @chancancode :)

**Comment by Kobzol** [comment]
Well, here goes nothing :) I merged the rewrite into master. @chancancode can you please make me an owner on `crates.io` using `cargo owner` so that I can publish this new version?

**Comment by chancancode** [comment]
Already did it yesterday! If it doesn’t work, lmk

**Comment by Kobzol** [comment]
When trying to run `cargo publish` it gives this error:
`error: api errors (status 200 OK): this crate exists but you don't seem to be an owner. If you believe this is a mistake, perhaps you need to accept an invitation to be an owner before publishing.`

And running `cargo owner --list` prints just `chancancode (Godfrey Chan)`.
I only created a working crates.io account today, so maybe it didn't work yesterday.

**Comment by chancancode** [comment]
I just tried it again and it said invitation sent. Do you see it in your inbox or some kind of notification thing on the web UI?

**Comment by Kobzol** [comment]
I accepted the invite and released a new version. Thank you very much :-)

**Comment by chancancode** [comment]
Thank you!

**Pull request merged**